### ISSUE TITLE
Option to ignore array index in `no-magic-number`?

### ISSUE THREAD

**Issue by mathieumg** [eslint/eslint#4370]
1. **The version of ESLint you are using.**

  1.9.0

2. **The problem you want to solve.**

  Accessing arrays using number literal indexes.

3. **Your take on the correct solution to problem.**

  Would it be welcome to introduce an option in the `no-magic-number` rule such that number literals used to access array indexes are not considered magic numbers?

1. **What you did.**

  ```js
  const example = matcher.exec(/myRegEx/);

  if (example) {
    const a = res[1];
    const b = res[2];
    const c = res[3];
    const d = res[4];
  }
  ```

  ESLint complains about `3` (c) and `4`(d), but it would also complain about `1` and `2` if they weren't default exceptions.
    
2. **What you would like to happen.**

  With the proper option, the `no-magic-number` would not flag syntax that follows the `array[#]` format where `#` is a number literal.

3. **What actually happened.**

  That currently is an error.

Thank you!

**Comment by eslintbot** [comment]
Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by lo1tuma** [comment]
Can you explain why those numbers shouldn’t be considered to be magic?

T

**Comment by mathieumg** [comment]
They're sequential array indexes to access RegEx match groups. I don't see myself creating constants such as `REGEX_MATCH_GROUP_X` with `X` going from 0 to a high number "just to be safe". I agree however that the number in your counterexample is indeed magic, and I imagine there's no way with the current static analysis tooling/approach/complexity to figure out if we're dealing with just any array or one that represents a match group.

That being said, I'd like to use `no-magic-numbers` in a more relaxed way, so I wouldn't mind opting in to an option that disables all array index checks.

Thank you!

**Comment by lo1tuma** [comment]
Sounds reasonable to me.

**Comment by nzakas** [comment]
Let's do it. @mathieumg want to pull together a PR?

**Comment by mathieumg** [comment]
Cool! Sure, I'm just not sure when I'll get around to it in the next month as I'll be very busy with school and finals. :P Perhaps on a calm Friday night!

**Comment by mathieumg** [comment]
I was thinking of looking into this in January, but thanks for handling it @cschuller ! I look forward to use the option. :)
